### PR TITLE
Enable use of records and bools for shared memory, remove ddt, add additional transpose tests

### PR DESCRIPTION
**Pull request by gmarkall**
This started out with a straightforward replacement of `ddt` with `unittest.subTest` - with the removal of support for Python 2, unittest subTests are available on all supported versions of Python. Since the use of `ddt` was quite minimal (only for the transpose test) I felt it was worth removing in favor of using `subTest` so that the vendored-in version of `ddt` can be dropped.

Whilst doing this I noticed that there weren't many tests for transpose, so I added a few more variants of sizes (around powers of 2, +/-1, where one or the other dimension is one of these, etc.) and noticed that record dtypes and bools were not permitted in shared arrays (as used by the transpose kernel) so I modified the implementation to allow them. This almost worked by simply allowing them through in `cudaimpl._generic_array` - however, NVVM requires shared memory variables to be power-of-2 aligned, which is true when using the sizes of all types in `types.number_domain` and booleans as their alignment, but not for records, which can be of arbitrary size. When the size is not a power of 2, it is rounded up to the next power of 2 to make it alignment to address this concern.

In summary, this PR:

* Adds support for record and bool dtypes for shared arrays and adds appropriate tests
* Adds additional variants of shapes and dtypes for the transpose tests
* Removes `ddt` and replaces it with the now built-in `subTest` function.

### CODE REVIEW

**Comment by seibert** [comment]
this passes internal CI

File: numba/cuda/cudaimpl.py [hunk 1]
```diff
@@ -641,7 +641,12 @@ def _generic_array(context, builder, shape, dtype, symbol_name, addrspace,
         # Create global variable in the requested address-space
         gvmem = lmod.add_global_variable(laryty, symbol_name, addrspace)
         # Specify alignment to avoid misalignment bug
-        gvmem.align = context.get_abi_sizeof(lldtype)
+        align = context.get_abi_sizeof(lldtype)
+        # Alignment is required to be a power of 2 for shared memory. If it is
+        # not a power of 2 (e.g. for a Record array) then round up accordingly.
+        if (align & (align - 1)) != 0:
+            align = 2 ** math.ceil(math.log2(align))
```
Comment (stuartarchibald) [inline numba/cuda/cudaimpl.py]: Is this `align = 1 << (align - 1 ).bit_length())` ?

File: numba/cuda/tests/cudapy/test_sm.py [hunk 2]
```diff
@@ -69,5 +74,56 @@ def costs_func(d_block_costs):
...
+
+            # One thread per block writes this block's chunk
+            if tx == 0:
+                for j in range(nthreads):
+                    y[bd * bx + j] = sm[j]
+
+        d_result = cuda.device_array_like(arr)
+        use_sm_chunk_copy[nblocks, nthreads](arr, d_result)
+        host_result = d_result.copy_to_host()
+        self.assertTrue(np.all(arr == host_result))
```
Comment (stuartarchibald) [inline numba/cuda/tests/cudapy/test_sm.py]: perhaps `self.assertEqual()` or `np.testing.assert_allclose`?, such that if it fails numerical comparison results are presented?

File: numba/cuda/tests/cudapy/test_transpose.py [hunk 3]
```diff
@@ -2,27 +2,64 @@
...
+
+    def test_transpose_record(self):
+        variants = ((2, 3), (16, 16), (16, 17), (17, 16), (14, 15), (15, 14),
+                    (14, 14))
+        for rows, cols in variants:
+            with self.subTest(rows=rows, cols=cols):
+                arr = np.recarray((rows, cols), dtype=recordwith2darray)
+                for x in range(rows):
+                    for y in range(cols):
+                        arr[x, y].i = x**2 + y
```
Comment (stuartarchibald) [inline numba/cuda/tests/cudapy/test_transpose.py]:
```suggestion
                        arr[x, y].i = x ** 2 + y
```

File: numba/cuda/tests/cudapy/test_transpose.py [hunk 4]
```diff
@@ -2,27 +2,64 @@
...
+                transposed = arr.T
+                d_arr = cuda.to_device(arr)
+                d_transposed = cuda.device_array_like(transposed)
+                transpose(d_arr, d_transposed)
+                host_transposed = d_transposed.copy_to_host()
+                self.assertTrue(np.all(transposed == host_transposed))
+
+    def test_transpose_bool(self):
+        variants = ((2, 3), (16, 16), (16, 17), (17, 16), (14, 15), (15, 14),
+                    (14, 14))
```
Comment (stuartarchibald) [inline numba/cuda/tests/cudapy/test_transpose.py]: Perhaps make these a class member? They are the same as the ones for `test_transpose_record`?

File: numba/cuda/tests/cudapy/test_transpose.py [hunk 5]
```diff
@@ -2,27 +2,64 @@
...
+                        arr[x, y].i = x**2 + y
+                        j = np.arange(3 * 2, dtype=np.float32)
+                        arr[x, y].j = j.reshape(3, 2) * x + y
+
+                transposed = arr.T
+                d_arr = cuda.to_device(arr)
+                d_transposed = cuda.device_array_like(transposed)
+                transpose(d_arr, d_transposed)
+                host_transposed = d_transposed.copy_to_host()
+                self.assertTrue(np.all(transposed == host_transposed))
```
Comment (stuartarchibald) [inline numba/cuda/tests/cudapy/test_transpose.py]: Same comment as previously about `self.assertEqual` etc.

File: numba/cuda/tests/cudapy/test_transpose.py [hunk 6]
```diff
@@ -2,27 +2,64 @@
...
+            with self.subTest(rows=rows, cols=cols):
+                arr = np.random.randint(2, size=(rows, cols), dtype=np.bool_)
+                transposed = arr.T
+
+                d_arr = cuda.to_device(arr)
+                d_transposed = cuda.device_array_like(transposed)
+                transpose(d_arr, d_transposed)
+
+                host_transposed = d_transposed.copy_to_host()
+                self.assertTrue(np.all(transposed == host_transposed))
```
Comment (stuartarchibald) [inline numba/cuda/tests/cudapy/test_transpose.py]: Same comment as previously about `self.assertEqual` etc.

File: numba/cuda/cudaimpl.py (hunk 1 above)
Comment (gmarkall) [reply numba/cuda/cudaimpl.py]: I believe so.

File: numba/cuda/tests/cudapy/test_sm.py (hunk 2 above)
Comment (gmarkall) [reply numba/cuda/tests/cudapy/test_sm.py]: Yes, that would be better. Since the data is a copy, I'd prefer an equality check rather than `assert_allclose`, because if they're only close then something strange happened.

File: numba/cuda/tests/cudapy/test_transpose.py (hunk 5 above)
Comment (gmarkall) [reply numba/cuda/tests/cudapy/test_transpose.py]: Same answer :-)

File: numba/cuda/tests/cudapy/test_transpose.py (hunk 4 above)
Comment (gmarkall) [reply numba/cuda/tests/cudapy/test_transpose.py]: Will do.

File: numba/cuda/tests/cudapy/test_transpose.py (hunk 6 above)
Comment (gmarkall) [reply numba/cuda/tests/cudapy/test_transpose.py]: Same answer again.

**Comment by seibert** [comment]
Passing internal and external CI again.  Ready for final review.

File: numba/cuda/cudaimpl.py [hunk 7]
```diff
@@ -645,7 +645,7 @@ def _generic_array(context, builder, shape, dtype, symbol_name, addrspace,
         # Alignment is required to be a power of 2 for shared memory. If it is
         # not a power of 2 (e.g. for a Record array) then round up accordingly.
         if (align & (align - 1)) != 0:
```
Comment (stuartarchibald) [inline numba/cuda/cudaimpl.py]: is this needed?

Comment (gmarkall) [reply numba/cuda/cudaimpl.py]: I don't think so (didn't fully understand your comment earlier) - I've removed it now.

**Comment by seibert** [comment]
rerun of CI passed
